### PR TITLE
x11: parse vendor and model out of xcb setup information

### PR DESCRIPTION
**Pull request by agx**

### CODE REVIEW

File: backend/x11/backend.c [hunk 1]
```diff
@@ -188,6 +188,23 @@ static void init_atom(struct wlr_x11_backend *x11, struct wlr_x11_atom *atom,
 	atom->reply = xcb_intern_atom_reply(x11->xcb_conn, atom->cookie, NULL);
 }
 
+
+static void parse_xcb_setup(struct wlr_output *output, xcb_connection_t *xcb_conn)
```
Comment (acrisci) [inline backend/x11/backend.c]: Brace should be on the same line as the function.

File: backend/x11/backend.c [hunk 2]
```diff
@@ -188,6 +188,23 @@ static void init_atom(struct wlr_x11_backend *x11, struct wlr_x11_atom *atom,
...
+	unsigned int length;
+
+	xcb_setup = xcb_get_setup (xcb_conn);
+	length = xcb_setup_vendor_length (xcb_setup);
+	if (length > sizeof(output->make)-1)
+		length = sizeof(output->make)-1;
+	memcpy (output->make, xcb_setup_vendor (xcb_setup), length);
+	output->make[length-1] = '\0';
+	sprintf(output->model, "%d.%d", xcb_setup->protocol_major_version,
+			xcb_setup->protocol_minor_version);
```
Comment (acrisci) [inline backend/x11/backend.c]: use `snprintf` here.

File: backend/x11/backend.c [hunk 3]
```diff
@@ -188,6 +188,23 @@ static void init_atom(struct wlr_x11_backend *x11, struct wlr_x11_atom *atom,
...
 
+
+static void parse_xcb_setup(struct wlr_output *output, xcb_connection_t *xcb_conn)
+{
+	const xcb_setup_t *xcb_setup;
+	unsigned int length;
+
+	xcb_setup = xcb_get_setup (xcb_conn);
+	length = xcb_setup_vendor_length (xcb_setup);
+	if (length > sizeof(output->make)-1)
```
Comment (acrisci) [inline backend/x11/backend.c]: always use braces on if statements.

File: backend/x11/backend.c [hunk 4]
```diff
@@ -188,6 +188,23 @@ static void init_atom(struct wlr_x11_backend *x11, struct wlr_x11_atom *atom,
 	atom->reply = xcb_intern_atom_reply(x11->xcb_conn, atom->cookie, NULL);
 }
 
+
+static void parse_xcb_setup(struct wlr_output *output, xcb_connection_t *xcb_conn)
+{
+	const xcb_setup_t *xcb_setup;
+	unsigned int length;
+
+	xcb_setup = xcb_get_setup (xcb_conn);
```
Comment (acrisci) [inline backend/x11/backend.c]: Don't put spaces before function parens.

**Comment by agx** [comment]
@acrisci thanks for review, all cleaned up now.

File: backend/x11/backend.c [hunk 5]
```diff
@@ -188,6 +188,24 @@ static void init_atom(struct wlr_x11_backend *x11, struct wlr_x11_atom *atom,
...
+static void parse_xcb_setup(struct wlr_output *output, xcb_connection_t *xcb_conn) {
+	const xcb_setup_t *xcb_setup;
+	unsigned int length;
+
+	xcb_setup = xcb_get_setup(xcb_conn);
+	length = xcb_setup_vendor_length (xcb_setup);
+	if (length > sizeof(output->make)-1) {
+		length = sizeof(output->make)-1;
+	}
+	memcpy (output->make, xcb_setup_vendor (xcb_setup), length);
```
Comment (ascent12) [inline backend/x11/backend.c]: I'd prefer that snprintf is used for this as well, which would actually remove the need for a lot of the preceding code:
```c
snprintf(output->make, sizeof(output->make), "%.*s",
    xcb_setup_vendor_length(xcb_setup),
    xcb_setup_vendor(xcb_setup));
```

File: backend/x11/backend.c [hunk 6]
```diff
@@ -188,6 +188,24 @@ static void init_atom(struct wlr_x11_backend *x11, struct wlr_x11_atom *atom,
 	atom->reply = xcb_intern_atom_reply(x11->xcb_conn, atom->cookie, NULL);
 }
 
+
+static void parse_xcb_setup(struct wlr_output *output, xcb_connection_t *xcb_conn) {
+	const xcb_setup_t *xcb_setup;
+	unsigned int length;
+
+	xcb_setup = xcb_get_setup(xcb_conn);
+	length = xcb_setup_vendor_length (xcb_setup);
```
Comment (ascent12) [inline backend/x11/backend.c]: No space before the `(`

File: backend/x11/backend.c [hunk 7]
```diff
@@ -188,6 +188,24 @@ static void init_atom(struct wlr_x11_backend *x11, struct wlr_x11_atom *atom,
 	atom->reply = xcb_intern_atom_reply(x11->xcb_conn, atom->cookie, NULL);
 }
 
+
+static void parse_xcb_setup(struct wlr_output *output, xcb_connection_t *xcb_conn) {
+	const xcb_setup_t *xcb_setup;
+	unsigned int length;
```
Comment (ascent12) [inline backend/x11/backend.c]: We're not using C89, so putting your declarations at the front isn't necessary. You can just declare them you use use them.

File: backend/x11/backend.c (hunk 5 above)
Comment (Ongy) [reply backend/x11/backend.c]: Why not strncpy?

Comment (agx) [reply backend/x11/backend.c]: xcb_setup_vendor is not null necessarily null terminated.

Comment (martinetd) [reply backend/x11/backend.c]: That's what the %.*s of printf is about, it's safe :)

Comment (agx) [reply backend/x11/backend.c]: @martinetd but it could still outgrow output->make

Comment (agx) [reply backend/x11/backend.c]: @martinetd ...no it can't sorry for the noise.

Comment (Ongy) [reply backend/x11/backend.c]: That's why it's `s*n*printf` the output->make size is the second argument to the function.

The maximum string length is then given as format option.

**Comment by agx** [comment]
All addressed now. Thanks for the snprintf( , "%.*s", ) hint.

File: backend/x11/backend.c [hunk 8]
```diff
@@ -188,6 +188,18 @@ static void init_atom(struct wlr_x11_backend *x11, struct wlr_x11_atom *atom,
...
+static void parse_xcb_setup(struct wlr_output *output, xcb_connection_t *xcb_conn) {
+	const xcb_setup_t *xcb_setup = xcb_get_setup(xcb_conn);
+
+	snprintf(output->make, sizeof(output->make), "%.*s",
+			xcb_setup_vendor_length(xcb_setup),
+			xcb_setup_vendor(xcb_setup));
+	snprintf(output->model, sizeof(output->model), "%"PRIu16".%"PRIu16,
+			xcb_setup->protocol_major_version,
+			xcb_setup->protocol_minor_version);
+
```
Comment (SirCmpwn) [inline backend/x11/backend.c]: style: drop this empty line

**Comment by agx** [comment]
Empty line removed.

**Pull request merged**